### PR TITLE
[Testcase Refactoring]Make test_sharded_optim.py device-generic with hw_classification label

### PR DESCRIPTION
**Pull request by youlituao**
## Summary

`test/distributed/_shard/sharded_optim/test_sharded_optim.py` tests
ShardedOptimizer / ShardedTensor semantics on distributed backends. Three
issues prevent out-of-tree backends from running these tests:

1. `@requires_accelerator_dist_backend(["nccl", "xccl"])` only allows
   NCCL and XCCL backends — out-of-tree PrivateUse1 backends are
   wrongly skipped.
2. Module-level `device_type` and `backend` global variables are used
   instead of the framework-injected `device` parameter — the
   instantiation framework cannot route to the correct device variant.
3. `TestShardedOptimizer` lacks `hw_classification` label and
   `instantiate_device_type_tests` instantiation — unlabeled tests are
   not filtered in CI pipelines.

This PR addresses all three issues by switching to capability-based
admission, removing module-level globals in favor of
`ShardedTensorTestBase.backend` (PR #194992), adding
`instantiate_device_type_tests` + spawn combination, and adding the
`hw_classification` label.

## Changes

- Replaced `@requires_accelerator_dist_backend(["nccl", "xccl"])` with
  `@requires_capabilities(Capability.distributed.backend)` in both test
  methods. This switches from a hardcoded backend name list to
  capability-based admission — device-agnostic, degrades to skipped
  (not failed) when the capability is not declared.
- Removed module-level `device_type` and `backend` global variables.
  Backend is now resolved at runtime via `ShardedTensorTestBase.backend`
  property (PR #194992) through `@with_comms(init_rpc=False)` default
  `backend=None`.
- Changed `@with_comms(init_rpc=False, backend=backend)` to
  `@with_comms(init_rpc=False)` in both test methods. With PR #194992,
  `with_comms` default `backend=None` resolves to `self.backend` at
  runtime via `dist.get_default_backend_for_device(self.device_type)`,
  where `self.device_type` is set by `instantiate_device_type_tests` per
  device variant.
- Added `device` parameter to both test methods (framework injection
  via `instantiate_device_type_tests`). `device_type` is derived via
  `torch.device(device).type` for sharding spec placement construction.
- Replaced `torch.rand([5, 10]).to(self.rank).requires_grad_()` with
  `torch.rand([5, 10], device=self.current_device, requires_grad=True)` —
  uses `self.current_device` property from `ShardedTensorTestBase` (PR #194992),
  which returns `torch.device(self.device_type, self.rank)`, keeping the
  per-rank device index and creating the tensor directly on the target device.
- `MyShardedLinear` simplified: removed `rank` and `device_type` constructor
  parameters and the redundant `if rank: .to(rank)` block (the subsequent
  `.to(self.current_device)` on the model handles device placement).
  `shard_parameter` still takes `device_type` for placement string construction.
- Added `instantiate_device_type_tests` import and file-end call with
  `except_for="cpu"` and `allow_xpu=True` — the test bodies require
  accelerators, so generating a CPU subclass is unnecessary.
  `allow_xpu=True` preserves the original XPU coverage from
  `@requires_accelerator_dist_backend(["nccl", "xccl"])`.
- Added `hw_classification = HardwareClassification.ACCELERATOR` class
  attribute.
- Removed the now-unused `requires_accelerator_dist_backend` import;
  added `Capability`, `requires_capabilities`, and
  `instantiate_device_type_tests` imports from `common_device_type`;
  added `HardwareClassification` import from `common_utils`.

## Not changed

- The test logic and assertions remain unchanged; only decorators,
  method signatures, device acquisition, and class attributes are
  modified.
- `MyShardedModel` is unchanged — it receives the sharding spec from
  the test method, not module-level globals.
- `@skip_if_lt_x_gpu(4)` is preserved (already hardware-agnostic).
- `ChunkShardingSpec` placement construction logic is unchanged (uses
  locally-derived `device_type` from `torch.device(device).type`).

## Depends on

This PR relies on the following upstream PRs (submitted but not yet merged):

- **#187650**: `skip_if_lt_x_gpu` hardware-agnostic implementation
  — provides accelerator-count check via `torch.accelerator`.
- **#190528**: `init_pg` hccl device binding
  — enables per-rank `set_device_index` for hccl backend.
- **#191919**: `Capability` mechanism + `requires_capabilities` decorator
  — provides the capability-based admission infrastructure.
- **#193080**: `PrivateUse1TestBase._capabilities()` override
  — declares `Capability.distributed.backend` for PrivateUse1 backends.
- **#194992**: `ShardedTensorTestBase.backend` property +
  `init_pg(backend=None)` + `with_comms(backend=None)`
  — enables runtime backend resolution from `self.device_type` instead
  of module-level global variables.

Without these, the test will be skipped on PrivateUse1 backends (graceful
degradation — skipped, not failed). The test PR can merge before the
common PRs due to this safe degradation behavior.

## Test plan

- `python test/distributed/_shard/sharded_optim/test_sharded_optim.py`
  on an out-of-tree accelerator backend (PrivateUse1, torch_npu
  2.13.0+git45fbeae on PyTorch 2.13.0a0+gitfad7424, Ascend 910B3,
  CANN 9.1.0-beta.1): `OK` — both tests pass
  (`TestShardedOptimizerPRIVATEUSE1` NPU variant, 4 processes each).
- `python test/distributed/_shard/sharded_optim/test_sharded_optim.py`
  on CPU (no accelerator): `NO TESTS RAN` — no device variants
  generated (`except_for="cpu"`), same as existing `only_for=("cuda",)`
  files on CPU-only machines.
- `python test/distributed/_shard/sharded_optim/test_sharded_optim.py`
  on CUDA (NVIDIA A100, 4 GPUs, CUDA 12.x): `OK` — both tests pass
  (`TestShardedOptimizerCUDA` CUDA variant). Confirms no regression on
  CUDA.

### CODE REVIEW

File: test/distributed/_shard/sharded_optim/test_sharded_optim.py [hunk 1]
```diff
@@ -118,8 +117,10 @@ def test_sharded_optim(self):
 
         before_update = deepcopy(sharded_optim.named_params)
 
-        inp = torch.rand(
-            [5, 10], device=torch.device(device_type, self.rank), requires_grad=True
+        inp = (
+            torch.rand([5, 10])
+            .to(torch.device(device_type, self.rank))
```
Comment (JamesD18) [inline test/distributed/_shard/sharded_optim/test_sharded_optim.py]: Use the current‑device defined in `ShardedTensorTestBase`.

Comment (youlituao) [reply test/distributed/_shard/sharded_optim/test_sharded_optim.py]: Updated. The input tensor is now created directly on the target device using self.device_type from ShardedTensorTestBase:

inp = torch.rand([5, 10], device=torch.device(self.device_type, self.rank), requires_grad=True)
This uses the class attribute self.device_type (set by setUpClass to the actual backend name, e.g. "npu"), keeps the per-rank device index, and avoids the CPU allocation + .to() transfer. The same pattern (torch.device(device_type, rank)) is applied in MyShardedLinear.__init__ for consistency.

**Review by JamesD18 — APPROVED** [review]
LGTM.Thanks for addressing the review comments.
